src/elf/gnu_hash.rs

```diff
@@ -1,143 +1,194 @@
-                GnuHash {
+                let chains_ptr = buckets_ptr.add(nbuckets as usize);
+                let chains = slice::from_raw_parts(chains_ptr, total_dynsyms - symindex as usize);
+                Ok(GnuHash {
                     nbuckets,
                     symindex,
+                    maskwords: maskwords - 1,
                     shift2,
-                    maskbits: mem::size_of::<usize>() as u32,
-                    bloomwords,
-                    hashvalues,
+                    bloom_filter,
                     buckets,
-                    maskwords_bitmask: ((maskwords as i32) - 1) as u32,
-                    symtab,
-                }
+                    chains,
+                })
             }
 
-            #[inline(always)]
-            fn lookup(&self,
-                      symbol: &str,
-                      hash: u32,
-                      strtab: &Strtab)
-                      -> Option<&sym::Sym> {
-                let mut idx = self.buckets[(hash % self.nbuckets) as usize] as usize;
-                // println!("lookup idx = buckets[hash % nbuckets] = {}", idx);
-                if idx == 0 {
+            /// Locate the hash chain, and corresponding hash value element.
+            fn lookup<Syms: IntoIterator<Item = sym::Sym>>(
+                &self,
+                symbol: &str,
+                hash: u32,
+                dynsyms: Syms,
+                dynstrtab: &Strtab,
+            ) -> Option<sym::Sym> {
+                const MASK_LOWEST_BIT: u32 = 0xffff_fffe; // == !1 in Rust
+                let bucket = self.buckets[(hash % self.nbuckets) as usize];
+                // Empty hash chain, symbol not present
+                if bucket < self.symindex {
                     return None;
                 }
-                let mut hash_idx = idx - self.symindex;
-                let hash = hash & !1;
-                // TODO: replace this with an iterator
-                loop {
-                    let symbol_ = &self.symtab[idx];
-                    let h2 = self.hashvalues[hash_idx];
-                    idx += 1;
-                    hash_idx += 1;
-                    let name = &strtab[symbol_.st_name as usize];
-                    // println!("{}: h2 0x{:x} resolves to: {}", i, h2, name);
-                    if hash == (h2 & !1) && name == symbol {
-                        // println!("lookup match for {} at: 0x{:x}", symbol, symbol_.st_value);
-                        return Some(symbol_);
+                // Walk the chain until the symbol is found or the chain is exhausted.
+                let chain_idx = bucket - self.symindex;
+                let hash = hash & MASK_LOWEST_BIT;
+                for (hash2, symb) in self
+                    .chains
+                    .iter()
+                    .skip(chain_idx as usize)
+                    .zip(dynsyms.into_iter().skip(bucket as usize))
```

Comment:
I think this iterator skip will be very inefficient compared with indexing like it used to. It defeats the purpose of using the hash in the first place.

Reply:
@lzutao What do you think?

Effectively, we're talking about this code:
```rust
 101   │                 loop {
 102   │                     let symbol_ = &self.symtab[idx];
 103   │                     let h2 = self.hashvalues[hash_idx];
 104   │                     idx += 1;
 105   │                     hash_idx += 1;
 106   │                     let name = &strtab[symbol_.st_name as usize];
 107   │                     // println!("{}: h2 0x{:x} resolves to: {}", i, h2, name);
 108   │                     if hash == (h2 & !1) && name == symbol {
 109   │                         // println!("lookup match for {} at: 0x{:x}", symbol, symbol_.st_value);
 110   │                         return Some(symbol_);
 111   │                     }
 112   │                     if h2 & 1 == 1 {
 113   │                         break;
 114   │                     } // end of chain
 115   │                 }
```

@philipc why do you think it will be less efficient? I'm not super familiar with skip, but I just assumed it would begin the indexing operations at the given idx, e.g., it expands to something like:

```
for i in skip_value..end
```
Perhaps this is incorrect?

Reply:
Fwiw, I used to benchmark the symbol resolution in dryad and it was getting close (but still somewhat far away) from the gnu linker. Didn't spend too much time, but `LD_DEBUG=statistics` is quite nice ballpark figures to test against. Unfortunately dryad is broken and i haven't had time to debug why the function call is now bad after there's been no code changes (except the rustc compiler) :(

Reply:
`skip(bucket as usize)` uses `Iterator::nth`, and the default implementation of `Iterator::nth` runs `Iterator::next` in a loop, which is far less efficient than indexing (as in `let symbol_ = &self.symtab[idx];`). We could probably fix this by implementing `Iterator::nth` for `elf::SymIterator`, but personally I don't think this function should accept an iterator, to avoid these kinds of footguns.

Reply:
K yes agreed, if that's how it is, skipping to index via iterating over everything (e.g., like accesing Nth element in a linked list) sucks; @lzutao sorry, needs to be fixed :( Did not know about this behavior of skip...